### PR TITLE
more flexible <port,core> mappings

### PR DESCRIPTION
**Pull request by yuyang0**
add `lcore_list` to port configuration.

### CODE REVIEW

File: lib/ff_config.h [hunk 1]
```diff
@@ -52,6 +53,9 @@ struct ff_port_cfg {
     char *broadcast;
     char *gateway;
     char *pcap;
+
+    int nb_lcores;
+    uint16_t lcore_list[DPDK_MAX_LCORE];
```
Comment (whl739) [inline lib/ff_config.h:58]: use `RTE_MAX_LCORE` instead.

File: lib/ff_config.c [hunk 2]
```diff
@@ -191,6 +192,90 @@ freebsd_conf_handler(struct ff_config *cfg, const char *section,
     return 1;
 }
 
+int uint16_cmp (const void * a, const void * b)
```
Comment (whl739) [inline lib/ff_config.c]: int
uint16_cmp (const void * a, const void * b)

File: lib/ff_config.c [hunk 3]
```diff
@@ -191,6 +192,90 @@ freebsd_conf_handler(struct ff_config *cfg, const char *section,
...
+parse_port_lcore_list(struct ff_port_cfg *cfg, const char *lcore_list)
+{
+    int i, j;
+    char input[4096];
+    char *tokens[128];
+    int nTokens = 0;
+    char *endptr;
+    uint16_t *cores = cfg->lcore_list;
+    int nr_cores = 0;
+    int max_cores = DPDK_MAX_LCORE;
```
Comment (whl739) [inline lib/ff_config.c]: use RTE_MAX_LCORE instead.

File: lib/ff_config.c [hunk 4]
```diff
@@ -191,6 +192,90 @@ freebsd_conf_handler(struct ff_config *cfg, const char *section,
...
+                if (nr_cores > max_cores) {
+                    fprintf(stderr, "too many cores\n");
+                    return 0;
+                }
+                cores[nr_cores++] = (uint16_t)j;
+            }
+        }
+    }
+    if (nr_cores <= 0) {
+        fprintf(stderr, "lcore list is empty\n");
```
Comment (whl739) [inline lib/ff_config.c]: just printf?

File: lib/ff_config.h (hunk 1 above)
Comment (yuyang0) [reply lib/ff_config.h:58]: if use `RTE_MAX_LCORE`, the `rte_config.h` must be included to `ff_config.h` and it will cause compile error.

Comment (whl739) [reply lib/ff_config.h:58]: well, some freebsd source files include the `ff_config.h`, cause the compiling issuse, so keep it this way.
And i'll look at the other codes tomorrow.

Comment (yuyang0) [reply lib/ff_config.h:58]: Actually we can make `lcore_list` as a pointer, and allocate memory when initialize `struct ff_port_cfg`, but I don't like this approach.

Comment (whl739) [reply lib/ff_config.h:58]: yeap.

**Comment by whl739** [comment]
There's a serious problem. See this case:
```
[dpdk]
lcore_mask=3
nb_ports=2
[port0]
lcore_list=0
[port1]
lcore_list=1
```
and start two processes:
```
./example/helloworld --conf=config.ini --proc-type=primary --proc-id=0
./example/helloworld --conf=config.ini --proc-type=secondary --proc-id=1
```
the second process will panic on `init_arp_ring`. Because the primary process didn't create arp_ring on port1.

This is mainly due to that my codes are high coupling, the primary process needs to initialize a lot of resources.
So this is a hard work, you have to modify the `init_arp_ring` and `init_port_start` too.

**Comment by whl739** [comment]
I almost forgot about `kni`, kni is only processed by the primary process, so you have to pay attention to it too.

**Pull request closed without merging**